### PR TITLE
Add BatchedTensorImpl

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38723 Add "batching rule" for torch.sum(tensor, dims)
* **#38424 Add BatchedTensorImpl**

On the way to adding initial vmap support, this is the implementation
for BatchedTensorImpl. Vmap (in future PRs) leverages Tensors backed by
BatchedTensorImpl to do its work.

For more context, here is an overview of the plan to add initial vmap support.
- [this PR] Add BatchedTensorImpl
- Add one or two batching rules
- Add vmap Python API
- Add "slow" for-loop fallbacks for out-of-place functions via
dispatcher fallback mechanism.
- Add batching rules for "view" functions
- Add "slow" for-loop fallbacks for in-place functions
- Miscellaneous handling for failure cases
- And more

Test Plan:
- `./build/bin/vmap_test`

Differential Revision: [D21640917](https://our.internmc.facebook.com/intern/diff/D21640917)

### CODE REVIEW

File: c10/core/DispatchKey.h [hunk 1]
```diff
@@ -150,6 +150,7 @@ enum class DispatchKey : uint8_t {
   // There are a number of alternative modes which may want to handle before
   // autograd; for example, error checking, tracing, profiling or vmap.  They
   // go here.
+  Vmap,
```
Comment (izdeby) [inline c10/core/DispatchKey.h:153]: nit: i would add 1-2 lines comment about the key it self.

File: aten/src/ATen/BatchedTensorImpl.h [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  }
+  int64_t level() const {
+    return level_;
+  }
+ private:
+  int64_t dim_;
+  int64_t level_;
+};
+
+using BatchDims = SmallVector<BatchDim, 5>;
```
Comment (izdeby) [inline aten/src/ATen/BatchedTensorImpl.h:28]: i would push '5' as a constant next to `kVmapMaxTensorDims`. Also, why 5?

File: aten/src/ATen/BatchedTensorImpl.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+using BatchDimsRef = ArrayRef<BatchDim>;
+
+struct TORCH_API BatchedTensorImpl : public c10::TensorImpl {
+  explicit BatchedTensorImpl(Tensor value, BatchDims bdims);
+
+  BatchDimsRef bdims() const { return bdims_; }
```
Comment (izdeby) [inline aten/src/ATen/BatchedTensorImpl.h:34]: does "Ref" part in the name implies that it should return reference?

File: aten/src/ATen/BatchedTensorImpl.h [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  Tensor value_;
+  BatchDims bdims_;
+};
+
+inline bool isBatched(const Tensor& tensor) {
+  return tensor.unsafeGetTensorImpl()->key_set().has(DispatchKey::Vmap);
+}
+
+inline BatchedTensorImpl* getBatched(Tensor tensor) {
+  return static_cast<BatchedTensorImpl*>(tensor.unsafeGetTensorImpl());
```
Comment (izdeby) [inline aten/src/ATen/BatchedTensorImpl.h:60]: `unsafeGetTensorImpl`, why is unsafe? sounds... unsafe

File: aten/src/ATen/BatchedTensorImpl.h [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+// Returns a bitset. If bit i is set, then that means dim i is a batchdim.
+inline std::bitset<kVmapMaxTensorDims> createIsBdimBitset(BatchDimsRef bdims) {
```
Comment (izdeby) [inline aten/src/ATen/BatchedTensorImpl.h:64]: nit: i would expand the name to be `createIsBdimBitset` to be `createBatchedDimBitset` cause with the first name is not obvious what is going on.

File: aten/src/ATen/BatchedTensorImpl.cpp [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  for (int64_t actual_dim = 0; actual_dim < kVmapMaxTensorDims; actual_dim++) {
+    if (is_bdim[actual_dim]) {
+      continue;
+    }
+    if (non_bdim_count == dim) {
+      return actual_dim;
+    }
+    non_bdim_count++;
+  }
+  TORCH_INTERNAL_ASSERT(false);
```
Comment (izdeby) [inline aten/src/ATen/BatchedTensorImpl.cpp:47]: Can you, please, add a comment on why we shouldn't get here? Or add an error message that would be obvious.

File: aten/src/ATen/BatchedTensorImpl.h (hunk 2 above)
Comment (zou3519) [reply aten/src/ATen/BatchedTensorImpl.h:28]: It's a hypothesis that users won't use more than 5 batch dimensions. In PyTorch we guarantee "best" performance on tensors of dimensionality less than or equal to 5 (TensorImpl's sizes/strides use `SmallVector<int, 5>`).

I'll toss the number into a constant as you suggested (y)

File: aten/src/ATen/BatchedTensorImpl.h (hunk 3 above)
Comment (zou3519) [reply aten/src/ATen/BatchedTensorImpl.h:34]: BatchDimsRef is typedef'ed to be an `ArrayRef<BatchDim>`. So yes, it is a reference

File: aten/src/ATen/BatchedTensorImpl.h (hunk 4 above)
Comment (zou3519) [reply aten/src/ATen/BatchedTensorImpl.h:60]: Unsafe because it returns the underlying TensorImpl. Users of our C++ API for instance should not be using it and the word "unsafe" deters them.

File: aten/src/ATen/BatchedTensorImpl.h (hunk 5 above)
Comment (zou3519) [reply aten/src/ATen/BatchedTensorImpl.h:64]: Sure

File: aten/src/ATen/BatchedTensorImpl.h [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+#include <ATen/ArrayRef.h>
+#include <ATen/SmallVector.h>
+#include <ATen/Tensor.h>
+
+namespace at {
+
+// We assume this in a few other places in the codebase,
+// but there isn't a centralized definition.
+constexpr int64_t kVmapMaxTensorDims = 64;
```
Comment (cpuhrsch) [inline aten/src/ATen/BatchedTensorImpl.h:13]: Why do you choose this constraint / number in particular?

File: aten/src/ATen/BatchedTensorImpl.h (hunk 4 above)
Comment (cpuhrsch) [reply aten/src/ATen/BatchedTensorImpl.h:64]: Also, in my experience, it even gives you the impl if the tensor is constant. So you could modify a constant Tensor.

File: c10/core/DispatchKey.h (hunk 1 above)
Comment (cpuhrsch) [reply c10/core/DispatchKey.h:156]: Why call it Vmap and not BatchedTensorKey or such?

File: aten/src/ATen/BatchedTensorImpl.h (hunk 7 above)
Comment (zou3519) [reply aten/src/ATen/BatchedTensorImpl.h:13]: If you create a tensor with more than 64 dims, a bunch of things in PyTorch will fail. Off of the top of my head:
- TensorIterator reduction operations will fail because they assume the input can have no more than 64 dims. (https://github.com/pytorch/pytorch/issues/23159)

**Comment by ezyang** [comment]
This PR is very pleasantly small, thank you!

File: aten/src/ATen/BatchedTensorImpl.h [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+// Store this number of elements of BatchDims on the stack. Most people will
+// probably use <= 5 nested vmaps, but adjust this number as necessary.
+constexpr int64_t kBatchDimsStackSize = 5;
+
+struct BatchDim {
```
Comment (ezyang) [inline aten/src/ATen/BatchedTensorImpl.h:19]: What is a batch dim? (Comment!)

File: aten/src/ATen/BatchedTensorImpl.h [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+  int64_t level_;
+};
+
+using BatchDims = SmallVector<BatchDim, kBatchDimsStackSize>;
+using BatchDimsRef = ArrayRef<BatchDim>;
+
+struct TORCH_API BatchedTensorImpl : public c10::TensorImpl {
+  explicit BatchedTensorImpl(Tensor value, BatchDims bdims);
+
+  BatchDimsRef bdims() const { return bdims_; }
```
Comment (ezyang) [inline aten/src/ATen/BatchedTensorImpl.h:38]: What's "bdims"? (Comment!)

File: aten/src/ATen/BatchedTensorImpl.h [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+  const Tensor& value() const { return value_; };
+
+  int64_t actualDim(int64_t dim, bool wrap_dim = true) const;
```
Comment (ezyang) [inline aten/src/ATen/BatchedTensorImpl.h:41]: What's actualDim? (comment!)

File: aten/src/ATen/BatchedTensorImpl.h [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+  // Override a bunch of methods inherited from TensorImpl.
```
Comment (ezyang) [inline aten/src/ATen/BatchedTensorImpl.h:43]: Noooooo more overrides :(   Is there a reason why BatchTensorImpl cannot just redundantly record the size/strides of its internal tensor?

File: aten/src/ATen/BatchedTensorImpl.h (hunk 4 above)
Comment (ezyang) [reply aten/src/ATen/BatchedTensorImpl.h:64]: It might be more in line with the prevailing naming if this one is also called getBatched. In fact, this is indeed unsafe, because it is not valid to call this on a non-batched tensor.

File: aten/src/ATen/BatchedTensorImpl.h (hunk 11 above)
Comment (zou3519) [reply aten/src/ATen/BatchedTensorImpl.h:43]: BatchedTensorImpl can handle sizes without overriding (it can set the inherited `sizes_` vector).

All of the overrides here are safety checks and internal asserts (e.g., I think that strides(), is_contiguous() should produce an error message but I'm not 100% sure). Another way to do the override is to add a branch for `isBatched(Tensor)` [here](https://github.com/pytorch/pytorch/blob/176174a68ba2d36b9a5aaef0943421682ecc66d4/aten/src/ATen/templates/TensorBody.h#L209-L211) and [here](https://github.com/pytorch/pytorch/blob/176174a68ba2d36b9a5aaef0943421682ecc66d4/aten/src/ATen/templates/TensorBody.h#L194-L196).

I'm fine with deleting the overrides and punting modifying `strides()` and `is_contiguous()` to the future after I've done some benchmarking to make sure adding the branch doesn't affect performance too much.

File: aten/src/ATen/BatchedTensorImpl.h [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+inline std::ostream& operator<<(std::ostream& out, const BatchDim& bdim) {
+  out << "(lvl=" << bdim.level() << ", dim=" << bdim.dim() << ")";
+  return out;
+}
+
+inline Tensor makeBatched(const Tensor& tensor, BatchDims bdims) {
+  TORCH_INTERNAL_ASSERT(!isBatched(tensor));
+  auto tensor_dim = tensor.dim();
+  TORCH_CHECK(
+      tensor_dim < kVmapMaxTensorDims,
```
Comment (ezyang) [inline aten/src/ATen/BatchedTensorImpl.h:85]: `<=`?

File: aten/src/ATen/BatchedTensorImpl.h (hunk 11 above)
Comment (ezyang) [reply aten/src/ATen/BatchedTensorImpl.h:43]: Oh it looks like you're overriding them to raise errors. OK

File: aten/src/ATen/BatchedTensorImpl.cpp [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+  // If we hit this assert, then that means
+  // `non_bdim_count` + #num_bdims > kVmapMaxTensorDims. We restrict the number
+  // of dims a BatchedTensorImpl can have to kVmapMaxTensorDims so this should
+  // never be hit.
+  TORCH_INTERNAL_ASSERT(false);
+}
+
+// The following are publically exposed as methods of Tensor
+IntArrayRef BatchedTensorImpl::strides() const {
+  TORCH_CHECK(false, "NYI: Getting tensor strides inside of vmap");
```
Comment (ezyang) [inline aten/src/ATen/BatchedTensorImpl.cpp:61]: Arguably you can give meaningful strides for a batch tensor. The idea is that you give the strides that, if correspondingly interpreted, would let you address the FIRST batch in the batched tensor impl (so the strides of the public dimensions, not the private one).

File: aten/src/ATen/BatchedTensorImpl.cpp [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+  }
+  auto is_bdim = createBatchDimBitset(bdims_);
+
+  // Example: assume dim = 3, and is_bdim = 10010011000...
+  // The 1's are batch dims and 0's are normal dims of the underlying value_ Tensor.
+  // actualDim gives us the index of `dim` in the `value_` Tensor, which is equivalent
+  // to asking "where does the 3rd (0-indexed) zero occur in the bitset?".
+  // The answer to that is index 5.
+  int64_t non_bdim_count = 0;
+  for (int64_t actual_dim = 0; actual_dim < kVmapMaxTensorDims; actual_dim++) {
```
Comment (ezyang) [inline aten/src/ATen/BatchedTensorImpl.cpp:43]: I feel like this should be implementable without a loop. Maybe https://stackoverflow.com/questions/7669057/find-nth-set-bit-in-an-int ? Don't know how portable the top liked answer is though.

Comment (ezyang) [reply aten/src/ATen/BatchedTensorImpl.cpp:43]: Although, since you're creating the bitset on the fly, I guess there isn't actually asymptotic improvement here :o)

File: aten/src/ATen/BatchedTensorImpl.cpp [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+  TORCH_INTERNAL_ASSERT(false, "Can't set_storage_offset for BatchedTensorImpl");
+}
+bool BatchedTensorImpl::has_storage() const {
+  TORCH_INTERNAL_ASSERT(false, "Can't query has_storage for BatchedTensorImpl");
+}
+
+Tensor addBatchDim(const Tensor& tensor, int64_t level, int64_t dim) {
+  if (!isBatched(tensor)) {
+    BatchDims bdims;
+    bdims.push_back({level, dim});
```
Comment (ezyang) [inline aten/src/ATen/BatchedTensorImpl.cpp:94]: c++ nit: `bdims.emplace_back(level, dim)`

File: aten/src/ATen/BatchedTensorImpl.cpp [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+    return at::detail::make_tensor<BatchedTensorImpl>(tensor, std::move(bdims));
+  }
+  const auto* batched = getBatched(tensor);
+  BatchDims new_bdims = { batched->bdims().begin(), batched->bdims().end() };
```
Comment (ezyang) [inline aten/src/ATen/BatchedTensorImpl.cpp:98]: C++ nit: It'd be better not to use braced initializer here, because it is easier to get confused with the initializer list form of this (in this case, it's unambiguous, but a reader might misinterpret as "take only the first and the last element of bdims", which is not the intent). So for example you could write instead: `BatchDims new_bdims(batched->bdims().begin(), batched->bdims().end())`. Maybe cc @smessmer for any further thoughts.

File: aten/src/ATen/BatchedTensorImpl.h (hunk 4 above)
Comment (ezyang) [reply aten/src/ATen/BatchedTensorImpl.h:64]: In this PR, you use getBatched in conjunction with isBatched. So an even safer form would be maybeGetBatched() that returns a nullptr if the result is not a batched tensor impl.

**Comment by ezyang** [comment]
At a high level, this PR seems very sparsely tested. I understand a lot of this will be tested more rigorously later. Probably you'll get the most utility from more heavy unit testing on the fiddly public/private batch dimensions calculations; esp with consideration for edge cases here.

**Comment by ezyang** [comment]
Looks promising! Looking forward to the next PRs in this set.

File: aten/src/ATen/BatchedTensorImpl.cpp (hunk 13 above)
Comment (zou3519) [reply aten/src/ATen/BatchedTensorImpl.cpp:61]: I agree that there is an argument that you can give meaningful strides. The reason why I am hesitant to provide it is because when users ask for strides, they are doing one of the following:
- they're about to grab the .data_ptr() and do some low-level things. vmap in its current state can't really support that.
- they're using the stride information to do some "performance tricks"; e.g. if you know what the strides of two matrices A, B are, it might be more performant to multiply them in some order (AB vs B^T A^T). The way vmap reorders dimensions could invalidate said performance tricks

File: aten/src/ATen/BatchedTensorImpl.cpp (hunk 14 above)
Comment (zou3519) [reply aten/src/ATen/BatchedTensorImpl.cpp:43]: I don't know what CPUs we support, but the instruction set (BMI2) used in that answer should be on Intel CPUs newer than 6/2013 and AMD from newer than 2015. I'll leave a comment in the code as a TODO to figure that out

File: aten/src/ATen/BatchedTensorImpl.h (hunk 4 above)
Comment (zou3519) [reply aten/src/ATen/BatchedTensorImpl.h:94]: I renamed `getBatched` to `unsafeGetBatched` and I created a `maybeGetBatched` variant (thanks for the suggestions!).

I have some downstream code that relies on `unsafeGetBatched` so I'd like to keep it around but we should be able to delete it in the future

**Review by ezyang — APPROVED** [review]

**Review by cpuhrsch — APPROVED** [review]
